AWSAppSyncClient/AppSyncSubscriptionWithSync.swift

```diff
@@ -0,0 +1,466 @@
+}
+
+public typealias DeltaSyncStatusCallback = ((_ currentState: DeltaSyncState) -> Void)
+
+public class SyncConfiguration {
+    
+    internal let seconds: Int
+    
+    internal var syncIntervalInSeconds: Int {
+        return seconds
+    }
+    
+    public init(seconds: Int) {
+        self.seconds = seconds
+    }
+    
+    // utility for setting default sync to 1 day
+    public class func defaultSyncConfiguration() -> SyncConfiguration {
+        return SyncConfiguration(seconds: 86400)
+    }
+}
+
+public typealias DeltaQueryResultHandler<Operation: GraphQLQuery> = (_ result: GraphQLResult<Operation.Data>?, _ transaction: ApolloStore.ReadWriteTransaction?, _ error: Error?) -> Void
+
+internal class AppSyncSubscriptionWithSync<Subscription: GraphQLSubscription, BaseQuery: GraphQLQuery, DeltaQuery: GraphQLQuery>: Cancellable {
+    
+    weak var appsyncClient: AWSAppSyncClient?
+    weak var subscriptionMetadataCache: AWSSubscriptionMetaDataCache?
+    var syncConfiguration: SyncConfiguration
+    var subscription: Subscription?
+    var baseQuery: BaseQuery?
+    var deltaQuery: DeltaQuery?
+    var subscriptionHandler: SubscriptionResultHandler<Subscription>?
+    var baseQueryHandler: OperationResultHandler<BaseQuery>?
+    var deltaQueryHandler: DeltaQueryResultHandler<DeltaQuery>?
+    var subscriptionWatcher: AWSAppSyncSubscriptionWatcher<Subscription>?
+    var userCancelledSubscription: Bool = false
+    var shouldQueueSubscriptionMessages: Bool = false
+    var subscriptionMessagesQueue: [(GraphQLResult<Subscription.Data>, Date)] = []
+    var isNetworkAvailable: Bool = true
+    var lastSyncTime: Date?
+    var deltaSyncOperationQueue: OperationQueue?
+    var subscriptionMessageDispatchQueue: DispatchQueue?
+    weak var handlerQueue: DispatchQueue?
+    var activeTimer: DispatchSourceTimer?
+    var deltaSyncStatusCallback: DeltaSyncStatusCallback?
+    var isSyncOperationSuccessful: Bool = false
+    var currentAttempt: Int = 0
+    
+    internal init(appsyncClient: AWSAppSyncClient,
+                  baseQuery: BaseQuery,
+                  deltaQuery: DeltaQuery,
+                  subscription: Subscription,
+                  baseQueryHandler: @escaping OperationResultHandler<BaseQuery>,
+                  deltaQueryHandler: @escaping DeltaQueryResultHandler<DeltaQuery>,
+                  subscriptionResultHandler: @escaping SubscriptionResultHandler<Subscription>,
+                  subscriptionMetadataCache: AWSSubscriptionMetaDataCache?,
+                  syncConfiguration: SyncConfiguration,
+                  handlerQueue: DispatchQueue) {
+        self.appsyncClient = appsyncClient
+        self.subscriptionMetadataCache = subscriptionMetadataCache
+        self.syncConfiguration = syncConfiguration
+        self.handlerQueue = handlerQueue
+        self.baseQuery = baseQuery
+        
+        if Subscription.operationString != "No-op" {
```

Comment:
I could not make them as optional since Swift compiler would not be able to figure the type of the generic parameters ([Ref](https://stackoverflow.com/questions/28458160/default-value-for-optional-generic-parameter-in-swift-function))

I will refactor `No-op` logic.